infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/StateTransferManager.java

```diff
@@ -0,0 +1,297 @@
+import org.corfudb.infrastructure.log.statetransfer.batch.TransferBatchRequest;
+import org.corfudb.infrastructure.log.statetransfer.batch.TransferBatchResponse;
+import org.corfudb.infrastructure.log.statetransfer.batch.TransferBatchResponse.TransferStatus;
+import org.corfudb.infrastructure.log.statetransfer.batchprocessor.StateTransferBatchProcessor;
+import org.corfudb.infrastructure.log.statetransfer.exceptions.TransferSegmentException;
+import org.corfudb.runtime.clients.LogUnitClient;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.LongStream;
+import java.util.stream.Stream;
+
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.TransferSegmentStatus.SegmentState.FAILED;
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.TransferSegmentStatus.SegmentState.NOT_TRANSFERRED;
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.TransferSegmentStatus.SegmentState.TRANSFERRED;
+
+/**
+ * A class responsible for managing a state transfer on the current node.
+ * It executes the state transfer for each non-transferred segment synchronously.
+ */
+@Slf4j
+@Builder
+public class StateTransferManager {
+
+    /**
+     * A log unit client to the current node.
+     */
+    @Getter
+    @NonNull
+    private final LogUnitClient logUnitClient;
+
+    /**
+     * A size of one batch of transfer.
+     */
+    @Getter
+    @NonNull
+    private final int batchSize;
+
+    /**
+     * A batch processor that transfers addresses one batch at a time.
+     */
+    @Getter
+    @NonNull
+    private final StateTransferBatchProcessor batchProcessor;
+
+
+    /**
+     * Given a range, return the addresses that are currently not present in the stream log.
+     *
+     * @param rangeStart Start address (inclusive).
+     * @param rangeEnd   End address (inclusive).
+     * @return A list of addresses, currently not present in the stream log.
+     */
+    @VisibleForTesting
+    ImmutableList<Long> getUnknownAddressesInRange(long rangeStart, long rangeEnd) {
+        Set<Long> knownAddresses = logUnitClient
+                .requestKnownAddresses(rangeStart, rangeEnd).join().getKnownAddresses();
+
+        return LongStream.range(rangeStart, rangeEnd + 1L)
+                .filter(address -> !knownAddresses.contains(address))
+                .boxed()
+                .collect(ImmutableList.toImmutableList());
+    }
+
+    /**
+     * Performs the state transfer for the current transfer segments and also
+     * updates their state as a result.
+     *
+     * @param transferSegments A list of the segment currently present in the system.
+     * @return A list with the updated transfer segments.
+     */
+    public ImmutableList<TransferSegment> handleTransfer(List<TransferSegment> transferSegments) {
+
+        return transferSegments.stream().map(segment -> {   // For each of the segments:
```

Comment:
This is true but in the scenarios where one segment fails and the other succeed it should work faster, since we retry the workflow and will just transfer the delta of addresses during the retry. In both cases the outcome will be the same, it's just in the case you suggested the retry will happen earlier in the case of epoch changes.